Python project template Vagrantfile 1.0 / Initial project docs (#1)
hdub-tech/vagrantfiles: python/README.md

  1. Pull down the [debian/bookworm64](https://app.vagrantup.com/debian/boxes/bookworm64) Vagrant box.
  2. Set hostname to `$project_name`.
  3. Create VM with the name `$project_name-debian12-bookworm` in VirtualBox.
  4. (root) Check for updated packages, upgrade system packages, and install the specified `$apt_packages` as well as the Python apt dependencies for required and optional modules.
  5. (vagrant) If `~/.pyenv` does not exist, install pyenv via the hated pipe to bash method. Add the pyenv setup config to `~/.pyenvrc`, and source that file from `~/.bashrc`.
  6. (vagrant) Install `$python_version` using [`pyenv`](https://github.com/pyenv/pyenv/blob/master/COMMANDS.md#pyenv-install) and set it as the local version for the project in `/vagrant`.
  7. (vagrant) Create a [venv](https://docs.python.org/3/library/venv.html#creating-virtual-environments) in `$venv_dir` with `$venv_prompt` and set it to activate on login.
  8. (vagrant) Run the user specified `$project_setup_script` and copy that script to `/vagrant/vagrant-project-setup.sh`.

---
## Variables

> This chart is the list of variables in the top of the Vagrantfile which you are encouraged to update based on your needs. To keep it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), rather than document default values here, please just see the corresponding [Vagrantfile](Vagrantfile).

Author: ```suggestion
This chart is the list of variables in the top of the Vagrantfile which you are encouraged to update based on your needs. To keep it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), rather than document default values here, please just see the corresponding [Vagrantfile](Vagrantfile#L6-L41).
```

| Variable | Type | Description |
| --- | --- | --- |
| `project_name` | String | Used as VM hostname and part of the VM name |

> | `python_version` | String | Any [pyenv supported version](https://github.com/pyenv/pyenv/tree/master/plugins/python-build/share/python-build) of Python. Major only, or major.minor versions supported |

Author: ```suggestion
| `python_version` | String | Any [pyenv supported version](https://github.com/pyenv/pyenv/tree/master/plugins/python-build/share/python-build) of Python, including the unlisted `major` or `major.minor` versions, like `3` or `3.12` |
```

| `create_new_project_venv` | boolean | Whether or not to create a new `.venv/vagrant-pyVERS` (`venv_dir`) with prompt `.venv $project_name-pyVERS` (`venv_prompt`) and activate it on login. If `false`, no venv actions will occur. If `true` and the `venv_dir` already exists, venv creation is skipped but it will be activated on login. |
| `apt_packages` | String (space separated)<sup>[1](#1)</sup> | The packaged which need to be installed in order to setup your project. You do NOT need to handle the Python dependencies here, just yours. |
| `project_setup_script` | Heredoc (Squiggly Quoted)<sup>[2](#2)</sup> | This is where your project specific setup script goes. Examples inlcude things like `pip install -e .` and `pip install tox` |

---
Footnotes:

>   * <a id=1>1</a> - I deliberately went with strings as opposed to an array to make this easier for users to populate. This also includes a few things that **_I_** like on everything and which you might not care for.

Author: ```suggestion
  * <a id=1>1</a> - I deliberately went with a string as opposed to an array to make this easier for users to populate. This also includes a few things that **_I_** like on everything and which you might not care for.
```
